### PR TITLE
refactor(monitor): Add new item circuit breaker on polling monitors

### PR DESCRIPTION
**Pull request by robzienert**
Adds a circuit breaker to each of the monitor pollers that will halt indexing a partition (a build master, usually) if the number of new cache items exceeds an upper boundary.

### CODE REVIEW

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/docker/DockerMonitor.groovy [hunk 1]
```diff
@@ -50,73 +64,72 @@ class DockerMonitor extends CommonPollingMonitor {
...
 
-            /* Not removing images from igor as we're seeing some reading issues in clouddriver
-            Observable.from(cachedImages).filter { String id ->
-                !(id in imageIds)
-            }.subscribe({ String imageId ->
-                log.info "Removing $imageId."
-                cache.remove(imageId)
-            }, {
-                log.error("Error: ${it.message}")
+            if (updateCache) {
```
Comment (dreynaud) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/docker/DockerMonitor.groovy]: I would pull the computation for updateCache in its own function, something like this:

```
def updateCache() {
    if (!imageId in cachedImages) {
        return true
    }

    if (trackDigests) {
        String lastDigest = cache.getLastDigest(image.account, image.registry, image.repository, image.tag)

        if (lastDigest != image.digest) {
            log.info("Updated tagged image: {}: {}. Digest changed from [$lastDigest] -> [$image.digest].", kv("account", image.account), kv("image", imageId))
            // If either is null, there was an error retrieving the manifest in this or the previous cache cycle.
            return image.digest != null && lastDigest != null
        }
    }

    return false
}
```

and then you don't need the variable I think:

```
if (updateCache()) {
    delta.add(new ImageDelta(imageId: imageId, image: image))
}
```

Comment (dreynaud) [reply igor-web/src/main/groovy/com/netflix/spinnaker/igor/docker/DockerMonitor.groovy]: I would also consider renaming the function, as it does not "update the cache". Maybe something like `shouldUpdateCache`?

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/docker/DockerMonitor.groovy [hunk 2]
```diff
@@ -50,73 +64,72 @@ class DockerMonitor extends CommonPollingMonitor {
...
 
     @Override
     String getName() {
         "dockerTagMonitor"
     }
 
-    void postEvent(EchoService echoService, List<String> cachedImagesForAccount, TaggedImage image, String imageId) {
+    void postEvent(List<String> cachedImagesForAccount, TaggedImage image, String imageId) {
+        if (!echoService.isPresent()) {
+            log.warn("Cannot send tagged image notification: Echo is not enabled")
```
Comment (dreynaud) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/docker/DockerMonitor.groovy]: would it make sense to publish a metric when this happens?

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/gitlabci/GitlabCiBuildMonitor.java [hunk 3]
```diff
@@ -44,64 +51,88 @@
...
+                    updatedBuilds.incrementAndGet();
+                    delta.add(new BuildDelta(branchedRepoSlug, project, pipeline, isPipelineRunning));
+                }
             }
+        });
 
-            log.info("Last poll took {} ms (master: {})", System.currentTimeMillis() - startTime, kv("master", master));
-        } catch (Exception e) {
-            log.error("Failed to obtain the list of projects", e);
+        if (delta.size() > 0) {
```
Comment (dreynaud) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/gitlabci/GitlabCiBuildMonitor.java]: `if (!delta.isEmpty())`

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/gitlabci/GitlabCiBuildMonitor.java [hunk 4]
```diff
@@ -44,64 +51,88 @@
...
+
+    @Override
+    protected void commitDelta(BuildPollingDelta delta) {
+        int ttl = buildCacheJobTTLSeconds();
+        final GitlabCiService gitlabCiService = (GitlabCiService) buildMasters.getMap().get(delta.master);
+
+        delta.items.parallelStream().forEach(item -> {
+            log.info("Build update [{}:{}:{}] [status:{}] [running:{}]", kv("master", delta.master), item.branchedRepoSlug, item.pipeline.getId(), item.pipeline.getStatus(), item.pipelineRunning);
+            buildCache.setLastBuild(delta.master, item.branchedRepoSlug, item.pipeline.getId(), item.pipelineRunning, ttl);
+            buildCache.setLastBuild(delta.master, item.project.getPathWithNamespace(), item.pipeline.getId(), item.pipelineRunning, ttl);
```
Comment (dreynaud) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/gitlabci/GitlabCiBuildMonitor.java]: why the double `setLastBuild` call?

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/IgorConfigurationProperties.groovy [hunk 5]
```diff
@@ -42,6 +42,19 @@ class IgorConfigurationProperties {
...
         final BuildProperties build = new BuildProperties()
+
+        @Canonical
+        static class PollingSafeguardProperties {
+            /**
+             * Defines the upper threshold for number of new cache items before a cache update cycle will be completely
+             * rejected. This is to protect against potentially re-indexing old caches due to downstream service errors.
+             * Once this threshold is tripped, API-driven resolution may be required.
+             */
+            int itemUpperThreshold = 10
```
Comment (dreynaud) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/IgorConfigurationProperties.groovy]: Aren't you worried that we can legitimately hit that limit on any given Jenkins master? Do we have metrics around the cache change rate as of today?

Comment (dreynaud) [reply igor-web/src/main/groovy/com/netflix/spinnaker/igor/IgorConfigurationProperties.groovy]: What does API-driven resolution mean?

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/polling/CommonPollingMonitor.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            T delta = generateDelta(ctx);
+
+            int upperThreshold = igorProperties.getSpinnaker().getPollingSafeguard().getItemUpperThreshold();
+            int deltaSize = delta.getItems().size();
+            if (deltaSize > upperThreshold) {
+                log.warn(
+                    "Number of items ({}) to cache exceeds upper threshold ({}) in {}:{}",
+                    deltaSize, upperThreshold, kv("monitor", monitorName), kv("partition", ctx.partitionName)
+                );
+                registry.gauge(itemsOverThresholdId.withTags("monitor", monitorName, "partition", ctx.partitionName)).set(deltaSize);
```
Comment (asher) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/polling/CommonPollingMonitor.java]: Explicitly set `itemsCachedId` gauge to 0 here too

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/gitlabci/GitlabCiBuildMonitor.java [hunk 7]
```diff
@@ -44,64 +51,88 @@
...
-    @Autowired(required = false)
-    private EchoService echoService;
-    @Autowired
-    private BuildCache buildCache;
-    @Autowired
-    private BuildMasters buildMasters;
+    private final BuildCache buildCache;
+    private final BuildMasters buildMasters;
+    private final GitlabCiProperties gitlabCiProperties;
+    private final Optional<EchoService> echoService;
```
[the remark this answers is not in the document]
Comment (robzienert) [reply igor-web/src/main/groovy/com/netflix/spinnaker/igor/gitlabci/GitlabCiBuildMonitor.java]: Autowired on properties is generally bad practice, so I refactored these classes to have a constructor, and autowired that instead.

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/gitlabci/GitlabCiBuildMonitor.java (hunk 4 above)
Comment (robzienert) [reply igor-web/src/main/groovy/com/netflix/spinnaker/igor/gitlabci/GitlabCiBuildMonitor.java]: From what I understand, GitlabCI tracks builds at both the project and branch level, so we need to track two different builds.

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/IgorConfigurationProperties.groovy (hunk 5 above)
Comment (ajordens) [reply igor-web/src/main/groovy/com/netflix/spinnaker/igor/IgorConfigurationProperties.groovy]: everything is fine ... re-index w/o sending notifications

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/build/BuildController.groovy [hunk 8]
```diff
@@ -209,7 +209,7 @@ class BuildController {
         }
     }
 
-    @RequestMapping(value = '/builds/properties/{buildNumber}/{fileName}/{master:.+}/**')
+    @RequestMapping(value = '/builds/igorProperties/{buildNumber}/{fileName}/{master:.+}/**')
```
Comment (ajordens) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/build/BuildController.groovy]: was s/properties/igorProperties on purpose?

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/docker/DockerMonitor.groovy (hunk 1 above)
Comment (ajordens) [reply igor-web/src/main/groovy/com/netflix/spinnaker/igor/docker/DockerMonitor.groovy]: 👍 to this suggestion

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/docker/DockerMonitor.groovy (hunk 2 above)
Comment (ajordens) [reply igor-web/src/main/groovy/com/netflix/spinnaker/igor/docker/DockerMonitor.groovy]: might not hurt

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/gitlabci/GitlabCiBuildMonitor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ajordens) [reply igor-web/src/main/groovy/com/netflix/spinnaker/igor/gitlabci/GitlabCiBuildMonitor.java]: 👍

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/gitlabci/GitlabCiBuildMonitor.java (hunk 4 above)
Comment (dreynaud) [reply igor-web/src/main/groovy/com/netflix/spinnaker/igor/gitlabci/GitlabCiBuildMonitor.java]: 👍

**Pull request merged**